### ISSUE TITLE
Feature request: Passing the scheduled start time to jobs

### ISSUE THREAD

**Issue by AdrianTeng** [agronholm/apscheduler#297]
Currently there are no way for the job to know what is the scheduled start time (only the current time by doing `datetime.now()`). This feature would be useful for time sensitive jobs (e.g. extra handling if the job is delayed)

**Comment by DeeVeX** [comment]
http://apscheduler.readthedocs.io/en/latest/modules/triggers/cron.html#module-apscheduler.triggers.cron
Isn't "start_date" what you are looking for?

**Comment by AdrianTeng** [comment]
@DeeVeX Nope. From what I understand from the docs, `start_date` is for setting the `CronTrigger`'s first trigger time. Whereas I want to know for each job, what is the scheduled trigger time, and passing that into the job

**Comment by DeeVeX** [comment]
Oh that would be nice.
I could use the same thing for a jobs last run to do cleanups!

**Comment by zhipcui** [comment]
Does anyone working on this?

**Comment by viiicky** [comment]
Yes, this feature is useful to have(I need this in my current project). Is there any work going on this? I am also open to submit PR for this if I get some light on which direction to move on.

Till then a workaround that I can think of is:
1. While adding job(`add_job()`) to scheduler pass an explicit identifier and store it somewhere.
2. When the job runs, fetch job(`get_job()`) from scheduler making use of the stored id and then do:
job.next_run_time - trigger's value
Something of that sort should work.

But I am not sure if it is full proof. And also, what would happen in the scenarios when using DB store and the scheduler goes down and jobs are missed and then the scheduler comes up. Would the calculation be still correct in that case

Either way waiting for this feature...

**Comment by viiicky** [comment]
Also I guess calling it `scheduled_time` instead of `scheduled_start_time` would be less confusing.
@AdrianTeng  @DeeVeX

**Comment by viiicky** [comment]
Mine use case by the way is explained below:

Suppose a job is scheduled to poll data from a server every 10 minutes.
Now the request that is made to server might have a `start_time` and `end_time` to be mentioned in the request being made. Something like: `fetch_messages_from_smtp_server(start_time, end_time)` or an equivalent REST API maybe.

So, I will `add_job` to scheduler mentioning the start time say 7th Sept 2018 12:00:00 UTC and an interval of 10 minutes.
And inside the job definition currently I would do something on these lines (psuedocode):
```
start_time = datetime.utcnow()
end_time = start_time + 10 mins
fetch_messages_from_smtp_server(start_time, end_time)
```
Now the catch here is, in an ideal world, I would expect the value of the call to `datetime.utcnow` to be as follows in the subsequent runs:
7th Sept 2018 12:00:00 UTC
7th Sept 2018 12:10:00 UTC
7th Sept 2018 12:20:00 UTC
.
.
.
But chances are that these values might be delayed for several reasons, in which case I might get undesired results from the server.

For example if the actual first run time is `7th Sept 2018 12:00:30 UTC` instead of `7th Sept 2018 12:00:00 UTC`, my request would be asking for all the messages between 7th Sept 2018 12:00:30 UTC to 7th Sept 2018 12:10:30 UTC in which case I would miss the messages for the first 30 secs and would possibly get some extra messages because of last 30 secs.

If there was a way where I could have accessed actual `scheduled_time`/`nominal_time` instead of just doing `utcnow` this would not have been a problem.

**Comment by richwifunds** [comment]
I need this feature too for measuring latency. Seems like it would be very easy to implement.

Just add { "scheduled_run_time": run_time } or something like that to **kwargs in line 125 of the BaseExecutor class - "retval = job.func(*job.args, **job.kwargs)".

Anyway, not sure the best way to do it, but since the value is right there, including it in the job.func call should be easy enough.

**Comment by vision57** [comment]
I think this is a widely required feature in such cron like scheduler, but is's quite strange that neither `apscheduler` nor `quartz` provide this.

**Comment by agronholm** [comment]
This will be implemented in v4.0 in such a manner that the target function can receive all sorts of information about the job.

**Comment by richwifunds** [comment]
Any update on this?

**Comment by richwifunds** [comment]
@agronholm  - has this been implemented and if not do you still plan to implement it and when?

**Comment by agronholm** [comment]
Preliminary work has been done but I'm mostly focusing on the highlight feature (data store sharing) now. It will take some time before 4.0 is in any usable state.

**Comment by 5uw1st** [comment]
I hope add a new event(JOB_ENENT_START) for this

**Comment by Dzeri96** [comment]
@agronholm Thanks for the quick response. Is there an ETA on v4? I've just started using this library.

**Comment by Dzeri96** [comment]
Alright, thanks for the update. I'm guessing that the onboarding process might be so difficult at the moment that it's not worth asking for contributions from your side.

**Comment by agronholm** [comment]
Implemented in v4.0.0a1 via the contextvar `apscheduler.current_job`.
